p2p/pqueue.go

```diff
@@ -0,0 +1,277 @@
+// make room for the incoming Envelope by dropping lower priority elements. If
+// there isn't sufficient capacity at lower priorities for the incoming Envelope,
+// it is dropped.
+//
+// After we attempt to enqueue the incoming Envelope, if the priority queue is
+// non-empty, we pop the top Envelope and send it on the dequeueCh.
+func (s *pqScheduler) process() {
+	defer s.done.Close()
+
+	for {
+		select {
+		case e := <-s.enqueueCh:
+			chIDStr := strconv.Itoa(int(e.channelID))
+			pqEnv := &pqEnvelope{
+				envelope:  e,
+				size:      uint(proto.Size(e.Message)),
+				priority:  s.chPriorities[e.channelID],
+				timestamp: time.Now().UTC(),
+			}
+
+			// enqueue
+
+			// Check if we have sufficient capacity to simply enqueue the incoming
+			// Envelope.
+			if s.size+pqEnv.size <= s.capacity {
+				// enqueue the incoming Envelope
+				s.push(pqEnv)
+			} else {
+				// There is not sufficient capacity to simply enqueue the incoming
+				// Envelope. So we have to attempt to make room for it by dropping lower
+				// priority Envelopes or drop the incoming Envelope otherwise.
+
+				// The cumulative size of all enqueue envelopes at the incoming envelope's
+				// priority or lower.
+				total := s.sizes[pqEnv.priority]
+
+				if total >= pqEnv.size {
+					// There is room for the incoming Envelope, so we drop as many lower
+					// priority Envelopes as we need to.
+					var (
+						canEnqueue bool
+						tmpSize    = s.size
+						i          = s.pq.Len() - 1
+					)
+
+					// Drop lower priority Envelopes until sufficient capacity exists for
+					// the incoming Envelope
+					for i >= 0 && !canEnqueue {
+						pqEnvTmp := s.pq.get(i)
+
+						if pqEnvTmp.priority < pqEnv.priority {
+							if tmpSize+pqEnv.size <= s.capacity {
+								canEnqueue = true
+							} else {
+								pqEnvTmpChIDStr := strconv.Itoa(int(pqEnvTmp.envelope.channelID))
+								s.metrics.PeerQueueDroppedMsgs.With("ch_id", pqEnvTmpChIDStr).Add(1)
+								s.logger.Debug(
+									"dropped envelope",
+									"ch_id", pqEnvTmpChIDStr,
+									"priority", pqEnvTmp.priority,
+									"msg_size", pqEnvTmp.size,
+									"capacity", s.capacity,
+								)
+
+								// dequeue/drop from the priority queue
+								heap.Remove(s.pq, pqEnvTmp.index)
+
+								// update the size tracker
+								tmpSize -= pqEnvTmp.size
+
+								// start from the end again
+								i = s.pq.Len() - 1
+							}
+						} else {
+							i--
+						}
+					}
+
+					// enqueue the incoming Envelope
+					s.push(pqEnv)
+				} else {
+					// There is not sufficient capacity to drop lower priority Envelopes,
+					// so we drop the incoming Envelope.
+					s.metrics.PeerQueueDroppedMsgs.With("ch_id", chIDStr).Add(1)
+					s.logger.Debug(
+						"dropped envelope",
+						"ch_id", chIDStr,
+						"priority", pqEnv.priority,
+						"msg_size", pqEnv.size,
+						"capacity", s.capacity,
+					)
+				}
+			}
+
+			// dequeue
+
+			for s.pq.Len() > 0 {
```

Comment:
Are we supposed to be popping a single envelope or all the envelopes in the queue?

Reply:
Yeah so here we pop off/dequeue and send on the dequeue channel as much as we can. This obviously creates a point of contention between enqueuing and dequeueing.

WDYT @tychoish ?

Reply:
Right so we send as many envelopes as possible until the channel blocks. Don't we want to do this before we enqueue a new envelope. Like perhaps the space that opens up when we dequeue allows for the new envelope to join the queue when beforehand we would've dropped it

Reply:
Or better still if the pq.Len() is 0 see if we can pop it on the dequeue channel straight away

Reply:
We send as much as there is, yes. If it blocks, it blocks, otherwise it sends.

I'm not sure the order matters as much as does the contention between enqueuing and dequeuing. We didn't see any performance degradation on idle or small load simple networks.

Reply:
Okay maybe you can't keep sending stuff to a buffered channel until it blocks. What I'm reading at the moment says that the code will panic with a fatal error

Reply:
We have a lot of places where things can theoretically block (because channels) in some [admittedly easy to imagine] pathological cases. 

All of these channels are buffered, which maybe gives a little bit of wiggle room, and maybe just makes the situation worse. I think the real point of contention is the bottleneck into/out of the heap (and it's mutex, likely,) which means, I think we can factor this anyway we want, and the heap is still the bottleneck. (e.g, having separate workers pulling incoming work and putting it on the heap.) 

It's potentially true that the the buffer of the outgoing channel should always be 0 or 1 to prevent messages from getting stale in line, but we could definitely have threads consuming work from the queue waiting for work? 

I can imagine ways to implement something with this functional component that would have less lock contention, or have "quicker" handling of messages, but I think the current implementation is definitely viable and given that the interface/implementations is entirely private, it seems worth waiting to see before [over]-engineering.

Reply:
Right, I'm all for simple first and [over]-engineering later but if this is the case can we make sure we document some of these thoughts so it's easier to pick up the conversation if we are to come back to it.